plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java

```diff
@@ -346,102 +419,285 @@ private void reorderCapellaDeactivationMessages(Scenario scenario,
+        } else {
+          // combined fragment found, check its contents
+          editElements(scenario,
+              ((org.polarsys.capella.scenario.editor.dsl.textualScenario.CombinedFragment) xtextElement).getBlock()
+                  .getBlockElements());
+          for (Operand operand : ((org.polarsys.capella.scenario.editor.dsl.textualScenario.CombinedFragment) xtextElement)
+              .getOperands()) {
+            editElements(scenario, operand.getBlock().getBlockElements());
+          }
         }
+      } else if (xtextElement instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) {
+        previousStateFragments.add(xtextElement);
+        editStateFragment(scenario,
+            (org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) xtextElement,
+            previousStateFragments);
       }
     }
   }
 
+  private static boolean foundMessagesBySourceAndTarget(EList<SequenceMessage> sequenceMessages, 
+      InstanceRole source, InstanceRole target, String messageName) {
+    if (source == null) {
+      return false;
+    }
+    List<SequenceMessage> msgsFilteredByNameTargetSource = sequenceMessages.stream()
+        .filter(x -> x.getName().equals(messageName))
+        .filter(x -> x.getSendingEnd().getCoveredInstanceRoles().get(0).getName().equals(source.getName()))
+        .filter(x -> x.getReceivingEnd().getCoveredInstanceRoles().get(0).getName().equals(target.getName()))
+        .collect(Collectors.toList());
+    return !msgsFilteredByNameTargetSource.isEmpty();
+  }
+
+  private static boolean hasReturn(Message elementFromXtext) {
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage)
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) elementFromXtext).getReturn() != null;
+    return false;
+  }
+  
+  protected static boolean hasExecution(Message elementFromXtext) {
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) {
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) elementFromXtext).getExecution() != null;
+    }
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.ArmTimerMessage) {
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.ArmTimerMessage) elementFromXtext).getExecution() != null;
+    }
+    return false;
+  }
+
+  private static void editStateFragment(Scenario scenario,
+      org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment xtextElement,
+      List<EObject> previousEditorStateFragments) {
+
+    InstanceRole instanceRole = EmbeddedEditorInstanceHelper.getInstanceRole(xtextElement.getTimeline());
+    EList<TimeLapse> ownedTimeLapses = scenario.getOwnedTimeLapses();
+
+    List<EObject> simillarStateFragments = previousEditorStateFragments.stream()
```

Comment:
This is an example of the complexity of the code when processing edge cases without any comment. I just have no idea why do we have to lood at similar StateFragments while creating new StateFragment here.